AWS ENI support

tgraf opened the pull request:
This PR adds AWS ENI IP allocation and datapath support. The details can be found in the individual commit messages and in the documentation commit.

Fixes: #6430

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8347)
<!-- Reviewable:end -->

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+	// Spec is the specification of the node
+	Spec NodeSpec `json:"spec"`
+
+	// Status it the status of the node
```
it --> is

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+	// +k8s:openapi-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec is the specification of the node
```
This is somewhat tautological.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+}
+
+// ENISpec is the ENI specification of a node. This specification is considered
+// by the cilium-operator to act as an IPAM operator and make ENI IPs available
```
make --> makes

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+// via the IPAMSpec section.
+//
+// The ENI specification can either be provided manually or the cilium agent
+// running ont he node can be instructed to create the CiliumNode custom
```
ont he --> on the

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+// by the cilium-operator to act as an IPAM operator and make ENI IPs available
+// via the IPAMSpec section.
+//
+// The ENI specification can either be provided manually or the cilium agent
```
What constitutes "manually"? This is not clear.

tgraf replied:
Changed it to:
> The ENI specification can either be provided explicitly by the user or the

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+	// InstanceType is the AWS EC2 instance type, e.g. "m5.large"
+	InstanceType string `json:"instance-type,omitempty"`
+
+	// MinAllocate is the minimum nimber of IPs that must be allocated when
```
nimber --> number

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+
+	// PreAllocate defines the number of IP addressses that must be
+	// available for allocation in the IPAMspec. It defines the buffer of
+	// addresses available immediately without requiring for the operator
```
requiring cilium-operator to get involved

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+type IPAMSpec struct {
+	// Available is the list of IPs available to the node for allocation.
+	// When an IP is used, the IP will remain on this list but will be
+	// added to Status.IPAM.InUse
```
I don't think the second sentence is necessary; it refers to another type of which `IPAMSpec` is not aware.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+	InUse map[string]AllocationIP `json:"used,omitempty"`
+}
+
+// AllocationIP is an IP available for allocation or already allocated
```
an IP which is available for allocation, or already has been allocated.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+
+// AllocationIP is an IP available for allocation or already allocated
+type AllocationIP struct {
+	// Owner is the owner of the IP, this field is set if the IP has been
```
... of the IP. This field is set if..

ianvernon commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,267 @@ type CiliumEndpointList struct {
+	ENIs map[string]ENI `json:"enis,omitempty"`
+}
+
+// ENI represents an AWS Elastic Network Interface
```
Provide a link to the specification of what an AWS Elastic Network Interface is here.

ianvernon commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -255,6 +261,32 @@ func createCEPCRD(clientset apiextensionsclient.Interface) error {
 	return createUpdateCRD(clientset, "v2.CiliumEndpoint", res)
 }
 
+// createNodECRD creates and updates the CiliumNode CRD. It should be called on
```
Nit: createNodeCRD

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+		TriggerFunc: store.refreshNodeTrigger,
+	})
+	if err != nil {
+		log.WithError(err).Fatal("Unable to initialize trigger")
```
This error is vague, make it more specific with an error message like "Unable to initialize CiliumNode synchronization trigger"

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if node, ok := obj.(*ciliumv2.CiliumNode); ok {
+					log.Infof("New CiliumNode %+v", node)
```
This will log at info level for every single `CiliumNode` that is added. Why not debug level?

tgraf replied:
Leftover, changed.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+					log.Infof("New CiliumNode %+v", node)
+					store.updateNodeResource(node)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", obj)
```
Will `%+v` give us the type of the resource as well? Adding `reflect.TypeOf(obj)` as well to this log message would give us more information if we ever hit this case.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+			DeleteFunc: func(obj interface{}) {
+				if node, ok := obj.(*ciliumv2.CiliumNode); ok {
+					log.Debugf("Deleted CiliumNode %+v", node)
+					store.updateNodeResource(nil)
```
I'm surprised that all of the Add, Delete, Update functions call the same function, `updateNodeResource` - the reasoning behind this is unclear.

tgraf replied:
In the delete case, it is invoking the function with `nil`. I have split it into two separate functions to make it clearer.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+				}
+			}
+		}
+	} else {
```
I don't understand why we would set `ownNode` to `nil` if we get a deletion event for any `CiliumNode`- add a comment explaining what is going on here.

tgraf replied:
The watcher is only watching the own node resource. I've renamed the functions to make it clearer.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+			break
+		}
+
+		log.Infof("Waiting for initial IP to become available in '%s' custom resource", nodeName)
```
This loops forever; should we have some `context.WithTimeout` to fail after some amount of time, say 5 minutes?

tgraf replied:
It will get killed by the liveness/readiness probe as the status API won't be served. Is that OK?

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+		&ciliumv2.CiliumNode{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
```
With regards to the lifecycle of `CiliumNodes`, how often are they created / deleted / updated? Is this usually a one-time resource that is not updated often? Are they updated frequently?

tgraf replied:
It is bound to the lifecycle of the k8s node. It is updated each time the IP allocation situation changes for that node. However, the watcher is only watching its own resource, not the resource of other nodes.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+		&ciliumv2.CiliumNode{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
```
We have metrics for other CRDs, we should add metrics here as well.

tgraf replied:
You are absolutely right, added.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+
+// refreshNodeTrigger is called to refresh the custom resource after taking the
+// configured rate limiting into account
+func (n *nodeStore) refreshNodeTrigger(reasons []string) {
```
`reasons` is never used - I expect this is so we can pass this as a function to `Trigger` - might be useful to mention this as a comment that this is intentional.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+	ciliumClient := k8s.CiliumClient()
+	switch {
+	case k8sCapabilities.UpdateStatus:
+		_, err = ciliumClient.CiliumV2().CiliumNodes("default").UpdateStatus(node)
```
Why the hardcoding of `default`?

tgraf replied:
Added constant in `pkg/k8s`. The resource is a global scope resource and not namespaced.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+// numAvailableIPs returns the total number of available IPs as known to the
+// node store
+func (n *nodeStore) numAvailableIPs() int {
+	if n.ownNode != nil {
```
Use `n.mutex.RLock`.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+	// mutex protects access to the allocated map
+	mutex lock.RWMutex
+
+	// allocated is a map of all allocated IPs
```
what does the string represent here?

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+	return allocator
+}
+
+func (a *crdAllocator) buildAllocationResult(ip net.IP, ipInfo *ciliumv2.AllocationIP) (result *AllocationResult, err error) {
```
Remove the `ipInfo` parameter if it isn't being used.

tgraf replied:
It was used in a follow-up commit. I've fixed it up and introduced the argument in the commit where it is first used.

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+	return
+}
+
+// Allocate will attempt to find the specified ip in the custom resourec and
```
resource

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+}
+
+// Allocate will attempt to find the specified ip in the custom resourec and
+// allocate it if it is available, if the IP is unavailable or already
```
... available. If the IP...

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+
+// Allocate will attempt to find the specified ip in the custom resourec and
+// allocate it if it is available, if the IP is unavailable or already
+// allocated, an error is return. The custom resource will be updated to
```
return --> returned

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+// Allocate will attempt to find the specified ip in the custom resourec and
+// allocate it if it is available, if the IP is unavailable or already
+// allocated, an error is return. The custom resource will be updated to
+// reflected the newly allocated IP.
```
reflected --> reflect

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+	a.store.refreshTrigger.TriggerWithReason(fmt.Sprintf("allocation of IP %s", ip.String()))
+}
+
+// AllocateNext allocates the next avalable IP as offered by the custom
```
available

ianvernon commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,435 @@
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+
+	allocs := map[string]string{}
```
Is the value in the map going to be updated later? If not, use `map[string]struct{}` if the value will always be empty.

tgraf replied:
Yes, it is used later:
```
                allocv4, st4 = ipam.IPv4Allocator.Dump()
                st4 = "IPv4: " + st4
                for ip := range allocv4 {
                        owner, _ := ipam.owner[ip]
                        // If owner is not available, report IP but leave owner empty
                        allocv4[ip] = owner
                }
```

I will create a good-first-issue to convert this into a type to hide this.

ianvernon commented on pkg/option/config.go:
```diff
@@ -501,6 +501,9 @@ const (
 	// IPAMHostScope is the value to select the hostscope IPAM plugin for option.IPAM
 	IPAMHostScope = "hostscope"
+
+	// IPAMCRD is the value to select the CRD-backed IPAM  plugin for option.IPAM
```
Remove the extra space in "IPAM  plugin"

ianvernon commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,121 @@
+					for _, nodeName := range nodeManager.GetNames() {
+						_, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+						switch {
+						case k8serrors.IsNotFound(err):
```
Is there an expected case where leftover nodes would accrue? I thought the watcher would take care of such cases? Given that assumption (which may be wrong on my part), I think we should log a warning if we have to garbage collect anything.

tgraf replied:
Correction from my previous statement: The combination of k8s node watcher plus the watcher of CiliumNode Add which will delete resources for which no k8s node can be found is sufficient to ensure removal of CIliumNode resources. I've removed the garbage collector.

ianvernon commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func deleteCiliumNode(name string) {
+	if err := ciliumK8sClient.CiliumV2().CiliumNodes("default").Delete(name, &metav1.DeleteOptions{}); err == nil {
```
Why only default?

tgraf replied:
Used constant k8s.DefaultNamespace to represent global scope of this resource.

ianvernon commented on operator/eni.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+func ciliumNodeDeleted(nodeName string) {
+	if nodeManager != nil {
```
Is there a case where the `nodeManager `would be `nil`? AFAICT we set its value synchronously before we invoke any of these functions in the watcher since `startENIAllocator` is invoked before `startSynchronizingCiliumNodes`.

tgraf replied:
Yes, when `--enable-eni=false` and `--synchronize-cilium-nodes=true`

aanm replied:
Why are we running `startSynchronizingCiliumNodes()` if `--enable-eni=false`?

ianvernon commented on daemon/daemon.go:
```diff
@@ -661,7 +664,25 @@ func createPrefixLengthCounter() *counter.PrefixLengthCounter {
+		}
+
+		if netConf.MTU != 0 {
+			configuredMTU = netConf.MTU
```
Maybe log that we are overriding `option.Config.MTU` with the MTU from the CNI configuration? These could conflict with one another from how I am reading this.

ianvernon commented on pkg/aws/eni/instances.go:
```diff
@@ -0,0 +1,164 @@
+	i.enis[eni.ID] = eni
+}
+
+// InstancesManager maintaines the list of instances. It must be kept up to
```
maintains

ianvernon commented on pkg/aws/eni/instances.go:
```diff
@@ -0,0 +1,164 @@
+// Resync fetches the list of EC2 instances and subnets and updates the local
+// cache in the instanceManager
+func (m *InstancesManager) Resync() {
+	//metricEniResync.Inc()
```
leftover?

tgraf replied:
Yes! Re-added metric.

ianvernon commented on pkg/aws/eni/limits.go:
```diff
@@ -0,0 +1,195 @@
+
+// limit contains limits for adapter count and addresses
+//
+// Source: github.com/lfyt/cni-ipvlan-vpc-k8s/
```
lfyt --> lyft

ianvernon replied:
Will these values ever change over time?

tgraf replied:
Possibly. It's planned to add a way to overwrite the limits manually. I'll create it as a good-first issue.

ianvernon commented on pkg/aws/metadata/metadata.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+func getMetadata(name string) (string, error) {
+	resp, err := http.Get("http://169.254.169.254/latest/meta-data/" + name)
```
What is this hardcoded IP?

tgraf replied:
It's the AWS metadata server IP. It's constant.

ianvernon commented on Documentation/concepts/datapath.rst:
```diff
@@ -0,0 +1,135 @@
+Egress
+~~~~~~
+
+1. The pod's network namespace contains default route which points to node's
```
contains a default route which points to the node's router IP...

ianvernon commented on Documentation/concepts/datapath.rst:
```diff
@@ -0,0 +1,135 @@
+
+1. The pod's network namespace contains default route which points to node's
+   router IP via the veth pair which is named ``eth0`` inside of the pod and
+   ``lxcXXXXXX`` in the host namespace. The router IP is allocated from ENI
```
The router IP is allocated from the ENI space, allowing for sending of ICMP errors from the router IP for Path MTU purposes.

ianvernon commented on Documentation/concepts/datapath.rst:
```diff
@@ -0,0 +1,135 @@
+   policies, implement load-balancing and provide networking features.
+
+3. An IP routing rule ensures that traffic from individual endpoints are using
+   a routing table specific to the ENI the endpoint IP allocated from:
```
a routing table specific to the ENI from which the endpoint IP was allocated:

ianvernon commented on Documentation/concepts/ipam/crd.rst:
```diff
@@ -0,0 +1,182 @@
+
+When this mode is enabled, each Cilium agent will start watching for a
+Kubernetes custom resource ``ciliumnodes.cilium.io`` with a name matching the
+Kubernetes node name the agent is running on.
```
the agent is running on --> on which the agent is running.

ianvernon commented on Documentation/concepts/ipam/eni.rst:
```diff
@@ -0,0 +1,366 @@
+  ``--auto-create-ciliumnode-resource`` or  set
+  ``auto-create-ciliumnode-resource: "true"`` in the ConfigMap.
+
+* It is generally a good idea to enable metrics in the Operator as well with
```
Perhaps note that Prometheus should be configured as well and link to that documentation here.

ianvernon reviewed (commented):
I think overall, the architecture of this is solid and well-documented. I'd probably have to immerse myself in the code by running it in a cluster and playing around to understand the nitty-gritty details given the size of the PR. 

Most of my comments are documentation / clarification about certain behaviors & operations. 

Are we marking this as GA for v1.6? Or beta?

joestringer commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,599 @@
+					store.updateLocalNodeResource(node)
+					owner.K8sEventProcessed("CiliumNode", "create", true)
+				} else {
+					log.Warningf("Unknown CiliumNode object type %s received: %+v", reflect.TypeOf(obj), obj)
```
nit: `log.WithFields(...)`... (And similarly for other user-visible warnings here)

tgraf replied:
It's actually intentionally, `WithFields()` does not expand "%+v" the same way. Is there a particular reason why fields are required for this highly unlikely warning that indicates a bug?

joestringer replied:
I was under the impression that we push for the `WithFields()` style so that log gatherers have a better time determining the same log message (similar to our top 5 thing in the CI), and that if the parameters are consistently named, you can do better correlation. The latter may be less applicable here given that the parameters are fairly unique.

tgraf replied:
I've changed all log messages to use fields except for the ones depending on `%+v` behavior.

joestringer commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,599 @@
+	// the custom resource with rate limiting
+	refreshTrigger *trigger.Trigger
+
+	// availableIPs is the cached version of all available IPs
```
```suggestion
	// availableIPs maps an address family to the size of the IP allocation pool for that family.
```

`availableIPs` seems to suggest it's to do with the number in the pool that can be allocated, but this appears to actually be the IP allocation pool size.

tgraf replied:
Renamed to `allocationPoolSize` to make it clear.

joestringer commented on daemon/daemon_main.go:
```diff
@@ -369,6 +368,9 @@ func init() {
 	option.BindEnv(option.BlacklistConflictingRoutes)
 
+	flags.Bool(option.AutoCreateCiliumNodeResource, defaults.AutoCreateCiliumNodeResource, "Automatically create CiliumNode resource for own node on startup")
+	option.BindEnv(option.AutoCreateCiliumNodeResource)
```
Any reason to have this as a separate option? Why not tie this into the IPAM in use, given that only some IPAM modules would make use of these resources (and, for that matter, those IPAMs require this to be set for correct operation)?

tgraf replied:
The main reason is to allow for users to provide their own custom resource via a script or similar when bootstrapping nodes. The PR allows for two ways to configure the ENI alllocation methods:

1. Via `--read-cni-conf`
2. By providing a custom resource directly and disabling auto-creation.

joestringer commented on daemon/daemon.go:
```diff
@@ -661,7 +664,26 @@ func createPrefixLengthCounter() *counter.PrefixLengthCounter {
+
 	bootstrapStats.daemonInit.Start()
+
+	if option.Config.ReadCNIConfiguration != "" {
```
Was this deliberately placed before option Config Validation?

tgraf replied:
No, good catch. Fixed.

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -506,6 +514,10 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
+// cmdDel is invoked on CNI DEL
+//
+// Note: ENI specific attributes do not need to be released as the ENIs and ENI
+// IPs are not released until the node termiantes and can be resused
```
```suggestion
// IPs can be reused and are not released until the node terminates.
```

joestringer commented on plugins/cilium-cni/eni.go:
```diff
@@ -0,0 +1,141 @@
+		return fmt.Errorf("ENI master interface MAC address is not set")
+	}
+
+	ifindex, err := prepareENI(ipam.MasterMac, int(conf.DeviceMTU))
```
IIRC in the regular Cilium operation, we configure MTU on devices for the maximum possible receive packet size (eg, from the internet); then we configure routes inside the endpoint with MTU to support the maximum anticipated packet size that the underlay will accept (commit 97a1a116241ab).

I guess that in this case, we don't expect to operate ENI with an overlay because the whole point is that ENI provides direct connectivity to all pods/nodes; so in that case we don't need to worry about configuring the MTU on the routes, as we can rely on the MTU on the devices. Is that in line with your understanding?

tgraf replied:
Correct. If a user configured encapsulation in combination with ENI IPAM then the ENI routes won't be used for any pod to pod traffic. I don't see much a reason to run in that mode though as direct routing mode has several advantages. Both modes will work fine though regarding MTU.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -587,36 +607,42 @@ func (m *IptablesManager) InstallRules(ifName string) error {
-				"-m", "comment", "--comment", "cilium host->cluster masquerade",
-				"-j", "SNAT", "--to-source", node.GetHostMasqueradeIPv4().String()}, false); err != nil {
-				return err
+			if option.Config.Tunnel != option.TunnelDisabled {
```
So we didn't need the masquerade rule for host->cluster in the direct routing mode, and this removes it?

Or is this now covered by something like the `cilium hostport cluster masquerade` rule?

tgraf replied:
That SNAT rule is required to ensure that reply traffic is sent back via the tunnel if we reach a remote node via the tunnel from a local source. If the source address selection picked a routable IP, the response may then be routed back directly if not SNATed.

tgraf replied:
The change could be done in a separate PR as well. It's needed for ENI to avoid having a rule pointing to `cilium_host` which is meaningless in ENI as well as to node depend on the allocation range.

joestringer commented on Documentation/requirements.txt:
```diff
@@ -8,7 +8,7 @@ imagesize==1.1.0
 jsonschema==2.6.0
 MarkupSafe==1.0
-Pygments==2.2.0
+Pygments==2.4.2
```
nit: the commit says you bumped to 2.4.3 but the code says 2.4.2

joestringer reviewed (approved):
Great commit messages, really helped to understand what's going on :tada: Most of my feedback is pretty minor.

Nit: I found the usage of `Available` with respect to IP address management/allocation to be a bit confusing. `Available` seems to be to do with the overall pool of IP addresses that can be allocated, but doesn't determine whether they are used/not. Ie a given IP can be both considered both "available" AND "used", which just means that the IP is in the pool of IPs to be allocated from, and it has been allocated.

There's quite a few places where warning logs are not using `WithFields()`, which means that potentially every instance of the warning will be unique. 

I skipped looking deeply into the main ENI commit, and the docs.

Perhaps in a follow-up PR, someone could jump in and add a CRV (Custom Resource Validator) and AdditionalPrinterColumns to improve pretty-printing of the new ciliumnodes resources? Might be a "good-first-issue".

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+	// get involved.
+	//
+	// +optional
+	PreAllocate int `json:"preallocate,omitempty"`
```
``json:"preallocate,omitempty"`` -> ``json:"pre-allocate,omitempty"``

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+	// allocated and are in use.
+	//
+	// +optional
+	InUse map[string]AllocationIP `json:"used,omitempty"`
```
```golang
InUse map[string]AllocationIP `json:"used,omitempty"`
```
json tag and field tag are different.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+// AwsSubnet stores information regarding an AWS subnet
+type AwsSubnet struct {
+	// ID is the ID of the subnet
+	ID string
```
missing json tag

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+	ID string
+
+	// CIDR is the CIDR range associated with the subnet
+	CIDR string
```
missing json tag

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+	PrimaryCIDR string
+
+	// CIDRs is the list of CIDR ranges associated with the VPC
+	CIDRs []string
```
missing json tag

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+	ID string
+
+	// PrimaryCIDR is the primary CIDR of the VPC
+	PrimaryCIDR string
```
missing json tag

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
+// AwsVPC stores information regarding an AWS VPC
+type AwsVPC struct {
+	/// ID is the ID of a VPC
+	ID string
```
missing json tag

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -72,6 +72,7 @@ cilium-agent [flags]
       --http-retry-timeout uint                    Time after which a forwarded but uncompleted request is retried (connection failures are retried immediately); defaults to 0 (never)
       --identity-change-grace-period duration      Time to wait before using new identity on endpoint identity change (default 5s)
       --install-iptables-rules                     Install base iptables rules for cilium to mainly interact with kube-proxy (and masquerading) (default true)
+      --ipam string                                Backend to use for IPAM (default "hostscope")
```
Shouldn't the default be denominated as "cilium"?

tgraf replied:
Are you suggesting to rename `"hostscope"` to `"cilium"`? I dont' care that much about how the option is called but `"cilium"` seems very unspecific.

aanm replied:
@tgraf yes. Otherwise it might be confusing for users as the "hostscope" name is extremely similar to the "host-local" cni plugin and it might give the impression that Cilium requires this "host-local" cni plugin to run.

tgraf replied:
What name do you propose? `"cilium"` is not meaningful.

aanm replied:
@tgraf Why not? Hmmm why not having `""` as default and write that "it defaults to internal IPAM if not specified."

tgraf replied:
That I'm fine with. I thought you meant to rename it to `"cilium"`

aanm commented on daemon/daemon.go:
```diff
@@ -767,6 +767,10 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
 			log.WithError(err).Fatal("Unable to initialize Kubernetes subsystem")
 		}
 
+		if err := k8s.RegisterCRDs(); err != nil {
```
Just a small note that this means `cilium-operator` won't register CRDs due this change

tgraf replied:
It's actually intentionally as we have an option to skip registration which was not respected by the operator. We can add registration via the operator but then we need to support that option as well.

aanm commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,444 @@
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if node, ok := obj.(*ciliumv2.CiliumNode); ok {
```
There's one other type that you need to try casting here.
```
					// Delete was not observed by the watcher but is
					// removed from kube-apiserver. This is the last
					// known state and the object no longer exists.
					deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
					if !ok {
						return
					}
```

tgraf replied:
As we are watching a single resource using a name selector, there is no need to cast at all. I've simplified it to treat any delete notification as the signal to forget about the resource.

aanm commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,444 @@
+			AddFunc: func(obj interface{}) {
+				if node, ok := obj.(*ciliumv2.CiliumNode); ok {
+					log.Debugf("New CiliumNode %+v", node)
+					store.updateLocalNodeResource(node)
```
Needs deepcopy

tgraf replied:
Moved the `DeepCopy()` done in `updateLocalNodeResource()` to here. I agree that it is more obvious.

aanm commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,444 @@
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if node, ok := newObj.(*ciliumv2.CiliumNode); ok {
+					log.Debugf("Updated CiliumNode %+v", node)
+					store.updateLocalNodeResource(node)
```
Needs deepcopy

aanm commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,444 @@
+	go ciliumNodeInformer.Run(wait.NeverStop)
+
+	log.WithField(fieldName, nodeName).Info("Waiting for CiliumNode custom resource to become available...")
+	if ok := cache.WaitForCacheSync(wait.NeverStop, ciliumNodeInformer.HasSynced); !ok {
```
Instead of `wait.NeverStop` maybe we should have a TTL? Otherwise we might never fatal

tgraf replied:
We will fatal. The loop below to wait for the IPs will block the API from coming up so the liveness probe will kill the pod eventually and it's clear from the log what is happening as the last log lines show that the agent was waiting for IPs to become available.

aanm commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,444 @@
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+
+	n.ownNode = node.DeepCopy()
```
Pass a deepcopy of the node already, it shouldn't be the responsibility of this function to do it.

aanm commented on pkg/ipam/crd.go:
```diff
@@ -0,0 +1,444 @@
+	// FIXME: This is currently using a brute-force method that can be
+	// optimized
+	for ip, ipInfo := range n.ownNode.Spec.IPAM.Pool {
+		if _, ok := allocated[ip]; !ok {
```
out of curiosity, why isn't the allocated IP map stored in the `nodeStore`?

tgraf replied:
It seemed more logical to store state in the allocator and make the nodeStore stateless. Either is fine I think. I'll open a good-first-issue to offer refactoring on this.

aanm commented on pkg/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,384 @@
+
+type metricsAPI interface {
+	ObserveEC2APICall(call, status string, duration float64)
+	IncEC2RateLimit(operation string)
```
This function does not seem related with metrics.

tgraf replied:
It is, it bumps the metric when the rate limiter kicked in.

aanm commented on pkg/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,384 @@
+	c.limiter.Wait(context.TODO())
+}
+
+// desribeNetworkInterfaces lists all ENIs
```
`describeNetworkInterfaces`

aanm commented on pkg/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,384 @@
+	}
+
+	for _, iface := range networkInterfaces {
+		err := parseAndAddENI(&iface, vpcs, subnets, instances)
```
Why don't you "Parse" only and return a `v2.ENI` + the error and add the `v2.ENI` to `instances` only if there's no errors? It will be easier to unit test the parsing logic alone.

tgraf replied:
Sure, changed.

aanm commented on pkg/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,384 @@
+
+// AttachNetworkInterface attaches a previously created ENI to an instance
+func (c *Client) AttachNetworkInterface(index int64, instanceID, eniID string) (string, error) {
+	attachReq := &ec2.AttachNetworkInterfaceInput{}
```
What's wrong initializing the `AttachNetworkInterfaceInput` with the fields?

```
attachReq := &ec2.AttachNetworkInterfaceInput{
   DeviceIndex: &index,
...
}
```

as it is done in the `ModifyNetworkInterface` function

aanm commented on pkg/aws/eni/instances.go:
```diff
@@ -0,0 +1,173 @@
+// instance is the minimal representation of an AWS instance as needed by the
+// ENI allocator
+type instance struct {
+	enis map[string]*v2.ENI
```
please add comment what is the key of this map.

aanm commented on pkg/aws/eni/instances.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+// GetENIs returns the list of ENIs associated with a particular instance
+func (m *InstancesManager) GetENIs(instanceID string) []*v2.ENI {
```
consider writing a private function that does this behavior so that `GetENI` can call that function otherwise it's odd that some public functions perform a RLock but not others `GetENI` vs `GetENIs`.

aanm commented on pkg/aws/eni/node.go:
```diff
@@ -0,0 +1,537 @@
+	return allocationNeeded
+}
+
+func (n *Node) resourceCopy() *v2.CiliumNode {
```
Consider making the function public as it is `RLocking`

tgraf replied:
I don't understand this comment

aanm replied:
@tgraf `resourceCopy() ` is a private method but it's also doing a RLock. Consider making the function public so that it is clear it can be accessed without worrying about the Lock being held or not or add a comment in the function saying that RLock is not required to be held.

tgraf replied:
Ah, now I get it. Done.

aanm commented on pkg/aws/eni/node.go:
```diff
@@ -0,0 +1,537 @@
+
+	// Always update the status first to ensure that the ENI information is
+	// synced for all addresses that are marked as available.
+	for retry := 0; retry < 2; retry++ {
```
Why only 2 attempts?

tgraf replied:
Added comment to explain

aanm commented on pkg/aws/eni/node_manager.go:
```diff
@@ -0,0 +1,237 @@
+						node.logger().WithError(err).Warning("Unable to resolve IP deficit of node")
+					}
+				} else {
+					node.logger().Warning("Node has disappeared while allocation request was queued")
```
node is nil here

aanm commented on pkg/aws/eni/node_manager.go:
```diff
@@ -0,0 +1,237 @@
+// Get returns the node with the given name
+func (n *NodeManager) Get(nodeName string) *Node {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
```
swap this line with the next one and remove the defer

aanm commented on pkg/aws/eni/node_manager.go:
```diff
@@ -0,0 +1,237 @@
+	return node
+}
+
+// byNeededIPs allows sorting of a slice of nodes by the number of addresses
```
You can remove lines 172-178

aanm commented on pkg/aws/eni/node_manager.go:
```diff
@@ -0,0 +1,237 @@
+		index++
+	}
+
+	sort.Sort(list)
```
And replace this with
`	sort.Slice(list, func(i, j int) bool { return list[i].getNeededAddresses() > list[j].getNeededAddresses() })`

aanm commented on pkg/aws/eni/node_manager.go:
```diff
@@ -0,0 +1,237 @@
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	list := make(byNeededIPs, len(n.nodes))
```
and `make(byNeededIPs, len(n.nodes))` with `make([]*Node, len(n.nodes))`

aanm commented on pkg/ipam/crd.go:
```diff
@@ -351,8 +366,50 @@ func newCRDAllocator(family Family) Allocator {
+	addr := subnetIP.To4()
+
+	// The gateway for a subnet and VPC is always x.x.x.1
+	// Ref: https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Route_Tables.html
```
If this is true then why isn't `return net.IPv4(addr[0], addr[1], addr[2], 1).String()` instead?

tgraf replied:
Comment was wrong, fixed

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -19,6 +19,7 @@ cilium-agent [flags]
       --agent-labels strings                       Additional labels to identify this agent
       --allow-localhost string                     Policy when to allow local stack to reach local endpoints { auto | always | policy } (default "auto")
       --annotate-k8s-node                          Annotate Kubernetes node (default true)
+      --auto-create-ciliumnode-resource            Automatically create CiliumNode resource for own node on startup
```
I'll defer to you but feels weird having `ciliumnode` and not `cilium-node`

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+var ciliumNodeStore cache.Store
+
+func convertToCiliumNode(obj interface{}) interface{} {
+	cnp, _ := obj.(*v2.CiliumNode)
```
Write a todo or something. Will the operator need to have all state and spec information about this node?

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+				if node, ok := obj.(*v2.CiliumNode); ok {
+					log.Debugf("Received new CiliumNode %+v", node)
+
+					_, err := k8s.Client().CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
```
Write a todo or something. The operator already has a list of all nodes so this Get is unnecessary.

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if node, ok := obj.(*v2.CiliumNode); ok {
```
Needs the 
```
					deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
					if !ok {
						return
					}
					// Delete was not observed by the watcher but is
					// removed from kube-apiserver. This is the last
					// known state and the object no longer exists.
```

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if node, ok := obj.(*v2.CiliumNode); ok {
+					log.Debugf("Received new CiliumNode %+v", node)
```
Remove this debug message. or change `node` to `node.Name`.

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if node, ok := newObj.(*v2.CiliumNode); ok {
+					log.Debugf("Received updated CiliumNode %+v", node)
```
Remove this debug message. or change `node` to `node.Name`.

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+			},
+			DeleteFunc: func(obj interface{}) {
+				if node, ok := obj.(*v2.CiliumNode); ok {
+					log.Debugf("CiliumNode got deleted %+v", node)
```
Remove this debug message. or change `node` to `node.Name`.

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+						log.WithField("name", node.Name).WithError(err).Warning("Unable to retrieve Kubernetes node")
+					}
+
+					ciliumNodeUpdated(node)
```
deepcopy node here

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if node, ok := newObj.(*v2.CiliumNode); ok {
+					log.Debugf("Received updated CiliumNode %+v", node)
+					ciliumNodeUpdated(node)
```
deepcopy node here

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+					log.Debugf("Received updated CiliumNode %+v", node)
+					ciliumNodeUpdated(node)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", newObj)
```
also add something like `"received type: %s", reflect.TypeOf(newObj)`

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+
+					ciliumNodeUpdated(node)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", obj)
```
also add something like `"received type: %s", reflect.TypeOf(newObj)`

aanm commented on operator/cilium_node.go:
```diff
@@ -0,0 +1,95 @@
+					log.Debugf("CiliumNode got deleted %+v", node)
+					ciliumNodeDeleted(node.Name)
+				} else {
+					log.Warningf("Unknown CiliumNode object received: %+v", obj)
```
also add something like `"received type: %s", reflect.TypeOf(newObj)`

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -547,3 +547,271 @@ type CiliumEndpointList struct {
 	Items []CiliumEndpoint `json:"items"`
 }
+
+// +genclient
```
```
// +genclient
// +genclient:nonNamespaced
```
you'll need to generate again autogenerated code.

aanm commented on operator/main.go:
```diff
@@ -100,9 +103,13 @@ func init() {
+	flags.String(option.IPAM, option.IPAMHostScope, "Backend to use for IPAM")
+	option.BindEnv(option.IPAM)
+	flags.BoolVar(&enableMetrics, "enable-metrics", false, "Enable Prometheus metrics")
+	flags.StringVar(&metricsAddress, "metrics-address", ":6942", "Address to serve Prometheus metrics")
```
why this port? please add this in the documentation.

tgraf replied:
Added documentation for all operator metrics

aanm commented on pkg/annotation/k8s.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 Authors of Cilium
+// Copyright 2018-2018 Authors of Cilium
```
it's ++year

aanm commented on pkg/datapath/linux/config.go:
```diff
@@ -271,7 +271,9 @@ func (l *linuxDatapath) writeTemplateConfig(fw *bufio.Writer, e datapath.Endpoin
 
-		fmt.Fprint(fw, "#define ENABLE_HOST_REDIRECT 1\n")
+		if option.Config.IPAM != option.IPAMENI {
+			fmt.Fprint(fw, "#define ENABLE_HOST_REDIRECT 1\n")
```
Doesn't this break ipvlan and/or flannel?

tgraf replied:
It only excludes this option for ENI mode. No change for IPVLAN or flannel mode.

aanm replied:
I read the other way around

aanm commented on plugins/cilium-cni/eni.go:
```diff
@@ -0,0 +1,141 @@
+)
+
+func prepareENI(mac string, mtu int) (index int, err error) {
+	var links []netlink.Link
```
mac = strings.ToLower(mac)

aanm commented on plugins/cilium-cni/eni.go:
```diff
@@ -0,0 +1,141 @@
+
+	// Route all traffic to the ENI address via the main routing table
+	if err := route.ReplaceRule(route.Rule{
+		Priority: 20,
```
Add comment, why 20?

tgraf replied:
Added commend

aanm commented on plugins/cilium-cni/eni.go:
```diff
@@ -0,0 +1,141 @@
+			// to the list of CIDRs configured for the VPC on which the
+			// endpoint has the IP on
+			if err := route.ReplaceRule(route.Rule{
+				Priority: 110,
```
Why 110?

aanm commented on plugins/cilium-cni/eni.go:
```diff
@@ -0,0 +1,141 @@
+	} else {
+		// Lookup a VPC specific table for all traffic from an endpoint
+		if err := route.ReplaceRule(route.Rule{
+			Priority: 110,
```
Same here for the 110

aanm commented on daemon/config.go:
```diff
@@ -82,6 +82,7 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 	changes += option.Config.Opts.ApplyValidated(om, changedOption, d)
 
 	log.WithField("count", changes).Debug("Applied changes to daemon's configuration")
+	option.Config.ConfigPatchMutex.Unlock()
```
Consider having this code in a different function that does the Lock/Unlock to avoid forgetting to do an `Unlock()` when we will add a new return.

tgraf replied:
Will create a good first issue to refactor this

aanm commented on pkg/option/config.go:
```diff
@@ -1068,6 +1083,20 @@ var (
+// IPv4NativeRoutingCIDR returns the native routing CIDR if configured
+func (c *DaemonConfig) IPv4NativeRoutingCIDR() *cidr.CIDR {
+	c.ConfigPatchMutex.RLock()
+	defer c.ConfigPatchMutex.RUnlock()
```
consider removing the defer

tgraf replied:
You really don't like these, don't you ;-)

aanm commented on Documentation/concepts/failure_behavior.rst:
```diff
@@ -0,0 +1,37 @@
+  their policy enforced but you will lose the ability to add additional
+  containers that belong to security identities which are unknown on the node;
+
+* If services are enabled, you will lose the ability to add additional services
```
Why? The allocation is local.

tgraf replied:
This is existing documentation. We can fix it up in a separate PR. This PR Is only splitting `concepts.rst` into individual files.

aanm commented on daemon/k8s_watcher.go:
```diff
@@ -1880,9 +1881,9 @@ func (d *Daemon) deleteK8sNodeV1(k8sNode *types.Node) error {
-func updateK8sEventMetric(scope string, action string, status bool) {
+// K8sEventProcessed is called to do metrics accounting for each processed
+// Kubernetes event
+func (d *Daemon) K8sEventProcessed(scope string, action string, status bool) {
```
Why is the Daemon a receiver?

tgraf replied:
Covered in comment below

aanm commented on pkg/ipam/crd.go:
```diff
@@ -175,25 +175,37 @@ func newNodeStore(nodeName string, owner Owner) *nodeStore {
+					valid = true
 					log.Debugf("New CiliumNode %+v", node)
 					store.updateLocalNodeResource(node)
+					owner.K8sEventProcessed("CiliumNode", "create", true)
```
Now I understand why the daemon is the receiver, maybe this `crd.go` should go under the `pkg/k8s` as well as the metrics function.

tgraf replied:
Covered in comment below

aanm commented on pkg/ipam/crd.go:
```diff
@@ -175,25 +175,37 @@ func newNodeStore(nodeName string, owner Owner) *nodeStore {
+					valid = true
 					log.Debugf("New CiliumNode %+v", node)
 					store.updateLocalNodeResource(node)
+					owner.K8sEventProcessed("CiliumNode", "create", true)
```
Now I understand why the daemon is the receiver, maybe this `crd.go` should go under the `pkg/k8s` as well as the metrics function.

tgraf replied:
Not completely opposed to it but it would require large refactoring of the IPAM code. We can do it separately.

aanm reviewed (changes requested):
Ping me directly if you don't understand any of my comments. Won't need my re-review after my changes are applied.

tgraf committed:
operator: Run operator in host networking mode

With the introduction of the ENI allocation capability, the operator must be
running in host networking mode as otherwise it would depends on its own
allocation capability in order to get scheduled.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Register CRDs earlier

We will have code during Daemon initialization rely on being able to access
custom resources.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
api: Expose masquerade status

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Grant RBAC access to CiliumNode resource

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
math: Add math package for IntMin() and IntMax()

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Add --ipam option to allow selecting IPAM backend

Default to existing hostscope plugin

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
api: Export additional IPAM information

The following additional information is exposed via the API:
 * IPAM mode
 * Gatway IP (optional)
 * Additional CIDRs to which the IP has access to (optional)
 * MAC of master interface (optional)

This is in preperation of upcoming ENI mode

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Support CILIUM_IPAM env in operator

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Provide additional IPAM allocation information

Building on the previous commit, this allows an IPAM plugin to provide the
additional allocation information.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
datapath: Extend ip routing rule support

This extends support to specify:
 - Priority
 - Mark mask
 - Source and destination filters

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
spanstat: Add Seconds() function

Simplifies use of spanstat for simple metrics

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
eni: Add ENI allocation logic

The AWS ENI allocator builds on top of the CRD-backed allocator. Each node a
creates a ciliumnodes.cilium.io custom resource matching the node name when
Cilium starts up for the first time on that node. It contacts the EC2 metadata
API to retrieve instance ID, instance type, and VPC information and populates
the custom resource with this information. ENI allocation parameters are
provided as agent configuration option and are passed into the custom resource
as well.

The architecture ensures that only a single operator communicates with the EC2
service API to avoid rate-limiting issues in bigger clusters. A pre-allocation
watermark allows to have IP addresses available for use on nodes at all time
without requiring to contact the EC2 API when a new pod is scheduled in the
cluster.

The Cilium operator listens for new ciliumnodes.cilium.io custom resources and
starts managing the IPAM aspect automatically. It scans the EC2 instances for
existing ENIs with associated IPs and makes them available via the
spec.ipam.available field. It will then constantly monitor the used IP
addresses in the status.ipam.used field and automatically create ENIs and
allocate more IPs as needed to meet the IP pre-allocation watermark. This
ensures that there are always IPs available

The selection of subnets to use for allocation as well as attachment of
security groups to new ENIs can be controlled separately for each node. This
makes it possible to hand out pod IPs with differing security groups on
individual nodes.

Configuration
-------------

* The Cilium agent and operator must be run with the option --ipam=eni or the
  option ipam: eni must be set in the ConfigMap. This will enable ENI
  allocation in both the node agent and operator.

Cache of ENIs and Subnets
-------------------------

The operator maintains a list of all EC2 ENIs and subnets associated with the
AWS account in a cache. For this purpose, the operator performs the following
two EC2 API operations:

* DescribeNetworkInterfaces
* DescribeSubnets

The cache is updated once per minute or after an IP allocation or ENI creation
has been performed. When triggered based on an allocation or creation, the
operation is performed at most once every 15 seconds.

Publication of available ENI IPs
--------------------------------

Following the update of the cache, all CiliumNode custom resources representing
nodes are updated to publish eventual new IPs that have become available.

In this process, all ENIs with an interface index greater than
spec.eni.first-interface-index are scanned for all available IPs. All IPs found
are added to spec.ipam.available. Each ENI meeting this criteria is also added
to status.eni.enis.

If this updated caused the custom resource to change, the custom resource is
updated using the Kubernetes API methods Update() and/or UpdateStatus() if
available.

Determination of ENI IP deficits
--------------------------------

The operator constantly monitors all nodes and detects deficits in available
ENI IP addresses. The check to recognize a deficit is performed on two
occasions:

* When a CiliumNode custom resource is updated
* All nodes are scanned in a regular interval (once per minute)

When determining whether a node has a deficit in IP addresses, the following
calculation is performed:

spec.eni.preallocate - (len(spec.ipam.available) - len(status.ipam.used))

Upon detection of a deficit, the node is added to the list of nodes which
require IP address allocation. When a deficit is detected using the interval
based scan, the allocation order of nodes is determined based on the severity
of the deficit, i.e. the node with the biggest deficit will be at the front of
the allocation queue.

The allocation queue is handled on demand but at most every 5 seconds.

IP Allocation
-------------

When performing IP allocation for a node with an address deficit, the operator
first looks at the ENIs which are already attached to the instance represented
by the CiliumNode resource. All ENIs with an interface index greater than
CiliumNode.Spec.ENI.FirstInterfaceIndex are considered for use.

The operator will then pick the first already allocated ENI which meets the
following criteria:

* The ENI has addresses associated which are not yet used or the number of
addresses associated with the ENI is lesser than the instance type specific
limit.

* The subnet associated with the ENI has IPs available for allocation

The following formula is used to determine how many IPs are allocated on the ENI:

min(AvailableOnSubnet, min(AvailableOnENI, NeededAddresses + spec.eni.max-above-watermark))

This means that the number of IPs allocated in a single allocation cycle can be
less than what is required to fulfill spec.eni.preallocate.

In order to allocate the IPs, the method AssignPrivateIpAddresses of the EC2
service API is called. When no more ENIs are available meeting the above
criteria, a new ENI is created.

ENI Creation
------------

As long as an instance type is capable allocating additional ENIs, ENIs are
allocated automatically based on demand.

When allocating an ENI, the first operation performed is to identify the best
subnet. This is done by searching through all subnets and finding a subnet that
matches the following criteria:

* The VPC ID of the subnet matches spec.eni.vpc-id
* The Availability Zone of the subnet matches spec.eni.availability-zone
* The subnet contains all tags as specified by spec.eni.subnet-tags
If multiple subnets match, the subnet with the most available addresses is
selected.

After selecting the ENI, the interface index is determine. For this purpose,
all existing ENIs are scanned and the first unused index greater than
spec.eni.first-interface-index is selected.

After determining the subnet and interface index, the ENI is created and
attached to the EC2 instance using the methods CreateNetworkInterface and
AttachNetworkInterface of the EC2 API.

The security groups attached to the ENI will be equivalent to
spec.eni.security-groups. The description will be in the following format:

"Cilium-CNI (<EC2 instance ID>)"

ENI Deletion Policy
-------------------

ENIs can be marked for deletion when the EC2 instance to which the ENI is
attached to is terminated. In order to enable this, the option
CiliumNode.Spec.ENI.DeleteOnTermination can be enabled. If enabled, the ENI is
modifying after creation using ModifyNetworkInterface to specify this deletion
policy.

Node Termination
----------------

When a node or instance terminates, the Kubernetes apiserver will send a node
deletion event. This event will be picked up by the operator and the operator
will delete the corresponding ciliumnodes.cilium.io custom resource.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Support setting ENI parameters via CNI configuration

Support specification of all ENI parameters via the CNI configuration file via
the option `--read-cni-conf`.

Also add support for a CNI chaining configuration in which case the
configuration is automatically derived from the Cilium specific section of the
chaining configuration

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cilium.io/v2: Generate k8s client code for new CiliumNode type

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cilium.io/v2: Add CiliumNode type definition

The CiliumNode custom resource is used to represent Cilium specific per node
configuration and status.

As a first use case, the CiliumNode custom resource is used to establish
communication between agents and the operator to allocate ENI IPs.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cni: Add ENI support

The ENI interface is automatically created and detected but needs to be brought
up and the MTU must be specified. The MTU is derived from the primary interface
by the agent and then inherited to all secondary ENIs.

Routing rules are set up to ensure that all traffic from endpoints are using
the ENI for egress which corresponds to the IP address assigned to the
endpoint.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: AWS ENI allocation ability

This adds operator support for ENI allocation. See the previous commits for the
ENI details. The operator will respect the `--ipam=eni` just like the agent. In
general, it makes sense to also enable --enable-metrics to get ENI specific
allocation metrics.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Add CRD-backed allocator

This adds an initial, minimal version of a CiliumNode CRD-backed IP allocator.
It is enabled by setting --ipam=crd and will hand out IPs based on the IPs made
available via the custom resource.

An outside operator or user can provide IPs via a map spec.ipam.avaiable. The
agent will use those IPs for allocation as they become available an update the
map status.ipam.used to list all used IPs.

This is a minimal version targeted for use by the operator for ENI support.
Further versions of this could become more sophisticated to also target manual
configuration use cases or integration into other IPAM systems.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Automatically create CiliumNode resource on startup

This adds an option --auto-create-ciliumnode-resource to automatically create a
CiliumNode custom resource on startup with the ENI parameters automatically
derived from the AWS metadata API.

The created custom resource is bound to the lifecycle of the Kubernetes node
resource so it gets automatically deleted when the node is removed.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cilium.io/v2: Register CiliumNode CRD

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
aws: Add metadata API package

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Split concepts section into multiple files

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Document ENI & CRD allocators

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Expose K8sEventReceived and K8sEventProcessed

Both are needed to perform CRD metric accounting for the IPAM CRD plugin.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Document cilium-operator metrics

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
eni: Support masquerading

This adds ENI support to the masquerading logic. For external SNAT, the
destination address exclusion is derived from the VPC's primary CIDR.

For now, the list of interfaces to masquerade on must be specified with
--egress-masquerade-interfaces with an interface selector. This is required
because filtering all local ENIs with a static rule is not possible yet. This
will be resolved once we switch to BPF based masquerading.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Bump pygments to version 2.4.2

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Add metrics accounting to CRD plugin

Signed-off-by: Thomas Graf <thomas@cilium.io>
